lib/active_record/associations/has_many_split_through_association.rb

```diff
@@ -4,28 +4,31 @@ module ActiveRecord
   module Associations
     class SplitAssociationScope < AssociationScope
       def scope(association)
+        # source of the through reflection
         reflection = association.reflection
+        #remove all previously set scope of passed in association
         scope = association.klass.unscoped
 
         chain = get_chain(reflection, association, scope.alias_tracker)
 
-        join_ids = [association.owner.id]
-        records = nil
-
         reverse_chain = chain.reverse
-        last_reflection = reverse_chain.last
-        reverse_chain.each do |refl|
-          records = refl.klass.unscoped.where(refl.join_keys.key => join_ids)
+        first_reflection = reverse_chain.shift
+        first_join_ids = [association.owner.id]
+        initial_values = [first_reflection, first_join_ids]
+
+        last_reflection, join_ids = reverse_chain.inject(initial_values) do |(reflection, join_ids), next_reflection|
```

Comment:
```suggestion
        last_reflection, last_join_ids = reverse_chain.inject(initial_values) do |(reflection, join_ids), next_reflection|
```

I recommend naming both variables last.